Snapshot failure

com.eclipsesource.workflow.glsp.server 0.0.1-SNAPSHOT FAILURE
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project com.eclipsesource.workflow.glsp.server: Compilation failure: Compilation failure: 
[ERROR] /Users/Experiment/Desktop/TestCoffee/coffee-editor/backend/plugins/com.eclipsesource.workflow.glsp.server/src/main/java/com/eclipsesource/workflow/glsp/server/model/WorkflowModelServerAccess.java:[34,35] package org.eclipse.emf.edit.domain does not exist
[ERROR] /Users/Experiment/Desktop/TestCoffee/coffee-editor/backend/plugins/com.eclipsesource.workflow.glsp.server/src/main/java/com/eclipsesource/workflow/glsp/server/model/WorkflowModelServerAccess.java:[35,35] package org.eclipse.emf.edit.domain does not exist
[ERROR] /Users/Experiment/Desktop/TestCoffee/coffee-editor/backend/plugins/com.eclipsesource.workflow.glsp.server/src/main/java/com/eclipsesource/workflow/glsp/server/model/WorkflowModelServerAccess.java:[66,17] cannot find symbol
[ERROR]   symbol:   class EditingDomain
[ERROR]   location: class com.eclipsesource.workflow.glsp.server.model.WorkflowModelServerAccess
[ERROR] /Users/Experiment/Desktop/TestCoffee/coffee-editor/backend/plugins/com.eclipsesource.workflow.glsp.server/src/main/java/com/eclipsesource/workflow/glsp/server/model/WorkflowModelServerAccess.java:[195,16] cannot find symbol
[ERROR]   symbol:   class EditingDomain
[ERROR]   location: class com.eclipsesource.workflow.glsp.server.model.WorkflowModelServerAccess
[ERROR] /Users/Experiment/Desktop/TestCoffee/coffee-editor/backend/plugins/com.eclipsesource.workflow.glsp.server/src/main/java/com/eclipsesource/workflow/glsp/server/handler/operation/AbstractCreateNodeHandler.java:[19,36] package org.eclipse.emf.edit.command does not exist
[ERROR] /Users/Experiment/Desktop/TestCoffee/coffee-editor/backend/plugins/com.eclipsesource.workflow.glsp.server/src/main/java/com/eclipsesource/workflow/glsp/server/handler/operation/AbstractCreateEdgeHandler.java:[17,36] package org.eclipse.emf.edit.command does not exist
[ERROR] /Users/Experiment/Desktop/TestCoffee/coffee-editor/backend/plugins/com.eclipsesource.workflow.glsp.server/src/main/java/com/eclipsesource/workflow/glsp/server/handler/operation/ApplyLabelEditOperationHandler.java:[17,36] package org.eclipse.emf.edit.command does not exist
[ERROR] /Users/Experiment/Desktop/TestCoffee/coffee-editor/backend/plugins/com.eclipsesource.workflow.glsp.server/src/main/java/com/eclipsesource/workflow/glsp/server/model/WorkflowModelServerSubscriptionListener.java:[25,35] package org.eclipse.emf.edit.domain does not exist
[ERROR] /Users/Experiment/Desktop/TestCoffee/coffee-editor/backend/plugins/com.eclipsesource.workflow.glsp.server/src/main/java/com/eclipsesource/workflow/glsp/server/model/WorkflowModelServerSubscriptionListener.java:[116,48] cannot find symbol
[ERROR]   symbol:   class EditingDomain
[ERROR]   location: class com.eclipsesource.workflow.glsp.server.model.WorkflowModelServerSubscriptionListener
[ERROR] /Users/Experiment/Desktop/TestCoffee/coffee-editor/backend/plugins/com.eclipsesource.workflow.glsp.server/src/main/java/com/eclipsesource/workflow/glsp/server/WorkflowGLSPModule.java:[14,37] package org.eclipse.emf.edit.provider does not exist
[ERROR] /Users/Experiment/Desktop/TestCoffee/coffee-editor/backend/plugins/com.eclipsesource.workflow.glsp.server/src/main/java/com/eclipsesource/workflow/glsp/server/model/WorkflowModelServerAccess.java:[75,42] cannot find symbol
[ERROR]   symbol:   class AdapterFactoryEditingDomain
[ERROR]   location: class com.eclipsesource.workflow.glsp.server.model.WorkflowModelServerAccess
[ERROR] /Users/Experiment/Desktop/TestCoffee/coffee-editor/backend/plugins/com.eclipsesource.workflow.glsp.server/src/main/java/com/eclipsesource/workflow/glsp/server/handler/operation/AbstractCreateNodeHandler.java:[51,38] cannot find symbol
[ERROR]   symbol:   variable AddCommand
[ERROR]   location: class com.eclipsesource.workflow.glsp.server.handler.operation.AbstractCreateNodeHandler
[ERROR] /Users/Experiment/Desktop/TestCoffee/coffee-editor/backend/plugins/com.eclipsesource.workflow.glsp.server/src/main/java/com/eclipsesource/workflow/glsp/server/handler/operation/AbstractCreateEdgeHandler.java:[50,38] cannot find symbol
[ERROR]   symbol:   variable AddCommand
[ERROR]   location: class com.eclipsesource.workflow.glsp.server.handler.operation.AbstractCreateEdgeHandler
[ERROR] /Users/Experiment/Desktop/TestCoffee/coffee-editor/backend/plugins/com.eclipsesource.workflow.glsp.server/src/main/java/com/eclipsesource/workflow/glsp/server/handler/operation/ApplyLabelEditOperationHandler.java:[45,38] cannot find symbol
[ERROR]   symbol:   variable SetCommand
[ERROR]   location: class com.eclipsesource.workflow.glsp.server.handler.operation.ApplyLabelEditOperationHandler
[ERROR] /Users/Experiment/Desktop/TestCoffee/coffee-editor/backend/plugins/com.eclipsesource.workflow.glsp.server/src/main/java/com/eclipsesource/workflow/glsp/server/model/WorkflowModelServerSubscriptionListener.java:[65,25] cannot find symbol
[ERROR]   symbol:   class EditingDomain
[ERROR]   location: class com.eclipsesource.workflow.glsp.server.model.WorkflowModelServerSubscriptionListener
[ERROR] /Users/Experiment/Desktop/TestCoffee/coffee-editor/backend/plugins/com.eclipsesource.workflow.glsp.server/src/main/java/com/eclipsesource/workflow/glsp/server/WorkflowGLSPModule.java:[125,59] cannot find symbol
[ERROR]   symbol:   class ComposedAdapterFactory
[ERROR]   location: class com.eclipsesource.workflow.glsp.server.WorkflowGLSPModule
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException

--- a/backend/plugins/com.eclipsesource.coffee.modelserver/pom.xml
+++ b/backend/plugins/com.eclipsesource.coffee.modelserver/pom.xml
@@ -33,7 +33,7 @@
 			</plugin>
 		</plugins>
 	</build>
-	
+
 	<dependencies>
 		<dependency>
 			<groupId>org.eclipse.emfcloud.modelserver</groupId>
@@ -41,12 +41,81 @@
 			<version>0.7.0-SNAPSHOT</version>
 		</dependency>
 		<dependency>
-			<groupId>com.eclipsesource.modelserver</groupId>
-			<artifactId>com.eclipsesource.modelserver.coffee.model</artifactId>
+			<groupId>org.eclipse.emfcloud.modelserver</groupId>
+			<artifactId>org.eclipse.emfcloud.modelserver.coffee.model</artifactId>
 			<version>0.7.0-SNAPSHOT</version>
 		</dependency>
+		<dependency>
+			<groupId>org.eclipse.emfcloud.modelserver</groupId>
+			<artifactId>org.eclipse.emfcloud.modelserver.common</artifactId>
+			<version>0.7.0-SNAPSHOT</version>
+		</dependency>
+		<dependency>
+			<groupId>org.eclipse.emfcloud.modelserver</groupId>
+			<artifactId>org.eclipse.emfcloud.modelserver.edit</artifactId>
+			<version>0.7.0-SNAPSHOT</version>
+		</dependency>
+		<dependency>
+			<groupId>org.eclipse.emfcloud.modelserver</groupId>
+			<artifactId>org.eclipse.emfcloud.modelserver.emf</artifactId>
+			<version>0.7.0-SNAPSHOT</version>
+		</dependency>
+		<dependency>
+			<groupId>org.eclipse.emfcloud.modelserver</groupId>
+			<artifactId>org.eclipse.emfcloud.modelserver.lib</artifactId>
+			<version>0.7.0-SNAPSHOT</version>
+		</dependency>
+		<dependency>
+			<groupId>log4j</groupId>
+			<artifactId>log4j</artifactId>
+			 <scope>provided</scope>
+		</dependency>
+		<dependency>
+			<groupId>org.slf4j</groupId>
+			<artifactId>slf4j-simple</artifactId>
+			<version>1.7.30</version>
+		</dependency>
+		<dependency>
+			<groupId>com.google.inject</groupId>
+			<artifactId>guice</artifactId>
+			 <scope>provided</scope>
+		</dependency>
+		<dependency>
+			<groupId>com.google.inject.extensions</groupId>
+			<artifactId>guice-multibindings</artifactId>
+			<version>3.0</version>
+		</dependency>
+		<dependency>
+			<groupId>org.eclipse.emf</groupId>
+			<artifactId>org.eclipse.emf.edit</artifactId>
+			<version>2.15.0</version>
+		</dependency>
+		<dependency>
+			<groupId>org.eclipse.emf</groupId>
+			<artifactId>org.eclipse.emf.ecore</artifactId>
+			 <scope>provided</scope>
+		</dependency>
+		<dependency>
+			<groupId>org.eclipse.emf</groupId>
+			<artifactId>org.eclipse.emf.ecore.xmi</artifactId>
+			 <scope>provided</scope>
+		</dependency>
+		<dependency>
+			<groupId>com.fasterxml.jackson.core</groupId>
+			<artifactId>jackson-databind</artifactId>
+			<version>2.10.0.pr1</version>
+		</dependency>
+		<dependency>
+			<groupId>org.emfjson</groupId>
+			<artifactId>emfjson-jackson</artifactId>
+			<version>1.2.0</version>
+		</dependency>
+		<dependency>
+			<groupId>com.google.guava</groupId>
+			<artifactId>guava</artifactId>
+			 <scope>provided</scope>
+		</dependency>
 	</dependencies>
-
 	<profiles>
 		<profile>
 			<id>fatjar</id>

--- a/backend/plugins/com.eclipsesource.workflow.glsp.server/pom.xml
+++ b/backend/plugins/com.eclipsesource.workflow.glsp.server/pom.xml
@@ -67,6 +67,21 @@
 			<artifactId>org.eclipse.elk.alg.layered</artifactId>
 			<version>0.5.0</version>
 		</dependency>
+		<dependency>
+			<groupId>org.eclipse.emf</groupId>
+			<artifactId>org.eclipse.emf.ecore</artifactId>
+			<scope>provided</scope>
+		</dependency>
+		<dependency>
+			<groupId>org.eclipse.emf</groupId>
+			<artifactId>org.eclipse.emf.edit</artifactId>
+			<version>2.15.0</version>
+		</dependency>
+		<dependency>
+			<groupId>com.fasterxml.jackson.core</groupId>
+			<artifactId>jackson-databind</artifactId>
+			<version>2.10.0.pr1</version>
+		</dependency>
 	</dependencies>
 	<build>
 		<plugins>

--- a/backend/releng/com.eclipsesource.coffee.target/com.eclipsesource.coffee.target.target
+++ b/backend/releng/com.eclipsesource.coffee.target/com.eclipsesource.coffee.target.target
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
 <?pde?>
 <!-- generated with https://github.com/eclipse-cbi/targetplatform-dsl -->
-<target name="Coffee 2020-09" sequenceNumber="1610375379">
+<target name="Coffee 2020-09" sequenceNumber="1612516697">
   <locations>
     <location includeMode="planner" includeAllPlatforms="false" includeSource="true" includeConfigurePhase="false" type="InstallableUnit">
       <unit id="com.google.gson" version="2.7.0.v20170129-0911"/>
@@ -34,7 +34,7 @@
     <location includeMode="planner" includeAllPlatforms="false" includeSource="true" includeConfigurePhase="false" type="InstallableUnit">
       <unit id="org.eclipse.emfcloud.modelserver.client" version="0.0.0"/>
       <unit id="org.eclipse.emfcloud.modelserver.common" version="0.0.0"/>
-      <repository location="https://download.eclipse.org/emfcloud/modelserver/p2/nightly/0.7/0.7.0.202101081124/"/>
+      <repository location="https://download.eclipse.org/emfcloud/modelserver/p2/nightly/0.7/0.7.0.202102041143/"/>
     </location>
     <location includeMode="planner" includeAllPlatforms="false" includeSource="true" includeConfigurePhase="false" type="InstallableUnit">
       <unit id="org.emfjson.jackson" version="0.0.0"/>

--- a/backend/releng/com.eclipsesource.coffee.target/com.eclipsesource.coffee.target.tpd
+++ b/backend/releng/com.eclipsesource.coffee.target/com.eclipsesource.coffee.target.tpd
@@ -38,7 +38,7 @@ location "http://download.eclipse.org/eclipse/updates/4.17/" eclipse-4.17 {
 location "http://download.eclipse.org/lsp4j/updates/releases/0.9.0" {
 	org.eclipse.lsp4j.sdk.feature.group lazy
 }
-location "https://download.eclipse.org/emfcloud/modelserver/p2/nightly/0.7/0.7.0.202101081124/" {
+location "https://download.eclipse.org/emfcloud/modelserver/p2/nightly/0.7/0.7.0.202102041143/" {
 	org.eclipse.emfcloud.modelserver.client lazy
 	org.eclipse.emfcloud.modelserver.common lazy
 }

--- a/web/.vscode/launch.json
+++ b/web/.vscode/launch.json
@@ -7,33 +7,7 @@
     {
       "type": "node",
       "request": "launch",
-      "name": "Start Browser Backend",
-      "program": "${workspaceRoot}/browser-app/src-gen/backend/main.js",
-      "args": [
-        "--loglevel=debug",
-        "--port=3000",
-        "--no-cluster",
-        "--WORKFLOW_NOTATION_LSP=5008",
-        "--app-project-path=${workspaceRoot}/browser-app",
-      ],
-      "env": {
-        "NODE_ENV": "development"
-      },
-      "sourceMaps": true,
-      "outFiles": [
-        "${workspaceRoot}/node_modules/@theia/*/lib/**/*.js",
-        "${workspaceRoot}/browser-app/lib/**/*.js",
-        "${workspaceRoot}/browser-app/src-gen/backend/*.js",
-        "${workspaceRoot}/packages/*/lib/**/*.js",
-      ],
-      "smartStep": true,
-      "internalConsoleOptions": "openOnSessionStart",
-      "outputCapture": "std"
-    },
-    {
-      "type": "node",
-      "request": "launch",
-      "name": "Debug Browser Backend",
+      "name": "Launch Browser Backend",
       "program": "${workspaceRoot}/browser-app/src-gen/backend/main.js",
       "args": [
         "--loglevel=debug",
@@ -43,6 +17,7 @@
         "--WF_LSP=5017",
         "--WF_ANALYZER=5083",
         "--app-project-path=${workspaceRoot}/browser-app",
+        "--plugins=${workspaceRoot}/browser-app/plugins"
       ],
       "env": {
         "NODE_ENV": "development"
@@ -52,7 +27,11 @@
         "${workspaceRoot}/node_modules/@theia/*/lib/**/*.js",
         "${workspaceRoot}/browser-app/lib/**/*.js",
         "${workspaceRoot}/browser-app/src-gen/backend/*.js",
-        "${workspaceRoot}/packages/*/lib/**/*.js",
+        "${workspaceRoot}/coffee-java-extension/lib/**/*.js",
+        "${workspaceRoot}/coffee-server/lib/**/*.js",
+        "${workspaceRoot}/coffee-workflow-analyzer/lib/**/*.js",
+        "${workspaceRoot}/coffee-workflow-analyzer-editors/lib/**/*.js",
+        "${workspaceRoot}/coffee-workflow-glsp-editor/lib/**/*.js"
       ],
       "smartStep": true,
       "internalConsoleOptions": "openOnSessionStart",

--- a/web/.vscode/tasks.json
+++ b/web/.vscode/tasks.json
@@ -1,42 +1,53 @@
 {
-    // See https://go.microsoft.com/fwlink/?LinkId=733558
-    // for the documentation about the tasks.json format
-    "version": "2.0.0",
-    "tasks": [
-        {
-            "label": "Coffee: Build all packages",
-            "type": "shell",
-            "group": {
-                "kind": "build",
-                "isDefault": true
-            },
-            "command": "yarn",
-            "presentation": {
-                "reveal": "always",
-                "panel": "new"
-            },
-            "problemMatcher": []
-        },
-        {
-            "label": "Coffee: Watch all packages",
-            "type": "shell",
-            "group": "build",
-            "command": "yarn watch",
-            "presentation": {
-                "reveal": "always",
-                "panel": "new"
-            },
-            "problemMatcher": []
-        },
-        {
-            "label": "Coffee: Start in Browser",
-            "type": "shell",
-            "command": "cd browser-app && yarn start",
-            "presentation": {
-                "reveal": "always",
-                "panel": "new"
-            },
-            "problemMatcher": []
-        }
-    ]
+  // See https://go.microsoft.com/fwlink/?LinkId=733558
+  // for the documentation about the tasks.json format
+  "version": "2.0.0",
+  "tasks": [
+    {
+      "label": "Coffee: Build all packages",
+      "type": "shell",
+      "group": {
+        "kind": "build",
+        "isDefault": true
+      },
+      "command": "yarn",
+      "presentation": {
+        "reveal": "always",
+        "panel": "new"
+      },
+      "problemMatcher": []
+    },
+    {
+      "label": "Coffee: Watch all packages",
+      "type": "shell",
+      "group": "build",
+      "command": "yarn watch",
+      "presentation": {
+        "reveal": "always",
+        "panel": "new"
+      },
+      "problemMatcher": []
+    },
+    {
+      "label": "Coffee: Start Browser Backend",
+      "type": "shell",
+      "command": "cd browser-app && yarn start",
+      "presentation": {
+        "reveal": "always",
+        "panel": "new"
+      },
+      "problemMatcher": []
+    },
+    {
+      "label": "Coffee: Open in Browser",
+      "type": "shell",
+      "group": "test",
+      "command": "google-chrome 127.0.0.1:3000",
+      "presentation": {
+        "reveal": "always",
+        "panel": "new"
+      },
+      "problemMatcher": []
+    }
+  ]
 }

--- a/web/coffee-editor-extension/src/browser/coffee-tree/coffee-tree-editor-widget.tsx
+++ b/web/coffee-editor-extension/src/browser/coffee-tree/coffee-tree-editor-widget.tsx
@@ -8,12 +8,11 @@
  *
  * SPDX-License-Identifier: EPL-2.0 OR MIT
  */
-import { ModelServerSubscriptionService } from '@eclipse-emfcloud/modelserver-theia/lib/browser';
 import {
     ModelServerClient,
-    ModelServerCommand,
     ModelServerCommandUtil,
     ModelServerReferenceDescription,
+    ModelServerSubscriptionService
 } from '@eclipse-emfcloud/modelserver-theia/lib/common';
 import {
     AddCommandProperty,
@@ -22,7 +21,7 @@ import {
     MasterTreeWidget,
     NavigatableTreeEditorOptions,
     NavigatableTreeEditorWidget,
-    TreeEditor,
+    TreeEditor
 } from '@eclipse-emfcloud/theia-tree-editor';
 import { Title, TreeNode, Widget } from '@theia/core/lib/browser';
 import { ILogger } from '@theia/core/lib/common';
@@ -34,286 +33,285 @@ import { CoffeeModel } from './coffee-model';
 
 @injectable()
 export class CoffeeTreeEditorWidget extends NavigatableTreeEditorWidget {
-  private delayedRefresh = false;
-  constructor(
-    @inject(MasterTreeWidget)
-    readonly treeWidget: MasterTreeWidget,
-    @inject(DetailFormWidget)
-    readonly formWidget: DetailFormWidget,
-    @inject(WorkspaceService)
-    readonly workspaceService: WorkspaceService,
-    @inject(ILogger) readonly logger: ILogger,
-    @inject(NavigatableTreeEditorOptions)
-    protected readonly options: NavigatableTreeEditorOptions,
-    @inject(ModelServerClient)
-    private readonly modelServerApi: ModelServerClient,
-    @inject(ModelServerSubscriptionService)
-    private readonly subscriptionService: ModelServerSubscriptionService
-  ) {
-      super(
-          treeWidget,
-          formWidget,
-          workspaceService,
-          logger,
-          CoffeeTreeEditorConstants.WIDGET_ID,
-          options
-      );
-      this.subscriptionService.onDirtyStateListener(dirtyState => {
-          this.dirty = dirtyState;
-          this.onDirtyChangedEmitter.fire();
-      });
-      this.subscriptionService.onFullUpdateListener(fullUpdate => {
-          this.instanceData = fullUpdate;
+    private delayedRefresh = false;
+    constructor(
+        @inject(MasterTreeWidget)
+        readonly treeWidget: MasterTreeWidget,
+        @inject(DetailFormWidget)
+        readonly formWidget: DetailFormWidget,
+        @inject(WorkspaceService)
+        readonly workspaceService: WorkspaceService,
+        @inject(ILogger) readonly logger: ILogger,
+        @inject(NavigatableTreeEditorOptions)
+        protected readonly options: NavigatableTreeEditorOptions,
+        @inject(ModelServerClient)
+        private readonly modelServerApi: ModelServerClient,
+        @inject(ModelServerSubscriptionService)
+        private readonly subscriptionService: ModelServerSubscriptionService
+    ) {
+        super(
+            treeWidget,
+            formWidget,
+            workspaceService,
+            logger,
+            CoffeeTreeEditorConstants.WIDGET_ID,
+            options
+        );
+        this.subscriptionService.onDirtyStateListener(modelServerMessage => {
+            this.dirty = modelServerMessage.data as boolean;
+            this.onDirtyChangedEmitter.fire();
+        });
+        this.subscriptionService.onFullUpdateListener(modelServerMessage => {
+            this.instanceData = modelServerMessage.data;
 
-          this.treeWidget
-              .setData({ error: false, data: this.instanceData })
-              .then(() => this.treeWidget.select(this.getOldSelectedPath()));
+            this.treeWidget
+                .setData({ error: false, data: this.instanceData })
+                .then(() => this.treeWidget.select(this.getOldSelectedPath()));
 
-          if (!this.isVisible) {
-              this.delayedRefresh = true;
-          }
-      });
-      this.subscriptionService.onIncrementalUpdateListener(incrementalUpdate => {
-          const command = incrementalUpdate as ModelServerCommand;
-          // the #/ marks the beginning of the actual path, but we also want the first slash removed so +3
-          const ownerPropIndexPath = command.owner.$ref
-              .substring(command.owner.$ref.indexOf('#/') + 3)
-              .split('/')
-              .filter(v => v.length !== 0)
-              .map(path => {
-                  const indexSplitPos = path.indexOf('.');
-                  // each property starts with an @ so we ignore it
-                  return {
-                      property: path.substring(1, indexSplitPos),
-                      index: path.substring(indexSplitPos + 1)
-                  };
-              });
-          let ownerNode;
-          if (ownerPropIndexPath.length !== 0) {
-              ownerNode = this.treeWidget.findNode(ownerPropIndexPath);
-          } else {
-              // TODO should be done in findNode
-              ownerNode = (this.treeWidget.model.root as TreeEditor.RootNode)
-                  .children[0];
-          }
-          const objectToModify =
-        ownerPropIndexPath.length === 0
-            ? this.instanceData
-            : ownerPropIndexPath.reduce(
-                (data, path) =>
-                    path.index === undefined
-                        ? data[path.property]
-                        : data[path.property][path.index],
-                this.instanceData
+            if (!this.isVisible) {
+                this.delayedRefresh = true;
+            }
+        });
+        this.subscriptionService.onIncrementalUpdateListener(modelServerMessage => {
+            const command = modelServerMessage.data;
+            // the #/ marks the beginning of the actual path, but we also want the first slash removed so +3
+            const ownerPropIndexPath = command.owner.$ref
+                .substring(command.owner.$ref.indexOf('#/') + 3)
+                .split('/')
+                .filter(v => v.length !== 0)
+                .map(path => {
+                    const indexSplitPos = path.indexOf('.');
+                    // each property starts with an @ so we ignore it
+                    return {
+                        property: path.substring(1, indexSplitPos),
+                        index: path.substring(indexSplitPos + 1)
+                    };
+                });
+            let ownerNode;
+            if (ownerPropIndexPath.length !== 0) {
+                ownerNode = this.treeWidget.findNode(ownerPropIndexPath);
+            } else {
+                // TODO should be done in findNode
+                ownerNode = (this.treeWidget.model.root as TreeEditor.RootNode)
+                    .children[0];
+            }
+            const objectToModify =
+                ownerPropIndexPath.length === 0
+                    ? this.instanceData
+                    : ownerPropIndexPath.reduce(
+                        (data, path) =>
+                            path.index === undefined
+                                ? data[path.property]
+                                : data[path.property][path.index],
+                        this.instanceData
+                    );
+            switch (command.type) {
+                case 'add': {
+                    if (!objectToModify[command.feature]) {
+                        objectToModify[command.feature] = [];
+                    }
+                    objectToModify[command.feature].push(...command.objectsToAdd);
+                    this.treeWidget.addChildren(
+                        ownerNode,
+                        command.objectsToAdd,
+                        command.feature
+                    );
+                    if (!this.isVisible) {
+                        this.delayedRefresh = true;
+                    }
+                    break;
+                }
+                case 'remove': {
+                    command.indices.forEach(i =>
+                        objectToModify[command.feature].splice(i, 1)
+                    );
+                    this.treeWidget.removeChildren(
+                        ownerNode,
+                        command.indices,
+                        command.feature
+                    );
+                    if (!this.isVisible) {
+                        this.delayedRefresh = true;
+                    }
+                    break;
+                }
+                case 'set': {
+                    // maybe we can directly manipulate the data?
+                    const data = clone(ownerNode.jsonforms.data);
+                    // FIXME handle array changes
+                    if (command.dataValues) {
+                        data[command.feature] = command.dataValues[0];
+                    } else {
+                        data[command.feature] = command.objectsToAdd[0];
+                    }
+                    this.treeWidget.updateDataForNode(ownerNode, data);
+                    if (!this.isVisible) {
+                        this.delayedRefresh = true;
+                    }
+                    if (this.selectedNode === ownerNode) {
+                        this.formWidget.setSelection(this.selectedNode);
+                    }
+                    break;
+                }
+                default: { /** */ }
+            }
+        });
+        this.modelServerApi.get(this.getModelIDToRequest()).then(response => {
+            if (response.statusCode === 200) {
+                if (isEqual(this.instanceData, response.body)) {
+                    return;
+                }
+                this.instanceData = response.body;
+                this.treeWidget
+                    .setData({ error: false, data: this.instanceData })
+                    .then(() => this.treeWidget.selectFirst());
+                return;
+            }
+            this.treeWidget.setData({ error: !!response.statusMessage });
+            this.renderError(
+                "An error occurred when requesting '" +
+                this.getModelIDToRequest() +
+                "' - Status " +
+                response.statusCode +
+                ' ' +
+                response.statusMessage
             );
-          switch (command.type) {
-              case 'add': {
-                  if (!objectToModify[command.feature]) {
-                      objectToModify[command.feature] = [];
-                  }
-                  objectToModify[command.feature].push(...command.objectsToAdd);
-                  this.treeWidget.addChildren(
-                      ownerNode,
-                      command.objectsToAdd,
-                      command.feature
-                  );
-                  if (!this.isVisible) {
-                      this.delayedRefresh = true;
-                  }
-                  break;
-              }
-              case 'remove': {
-                  command.indices.forEach(i =>
-                      objectToModify[command.feature].splice(i, 1)
-                  );
-                  this.treeWidget.removeChildren(
-                      ownerNode,
-                      command.indices,
-                      command.feature
-                  );
-                  if (!this.isVisible) {
-                      this.delayedRefresh = true;
-                  }
-                  break;
-              }
-              case 'set': {
-                  // maybe we can directly manipulate the data?
-                  const data = clone(ownerNode.jsonforms.data);
-                  // FIXME handle array changes
-                  if (command.dataValues) {
-                      data[command.feature] = command.dataValues[0];
-                  } else {
-                      data[command.feature] = command.objectsToAdd[0];
-                  }
-                  this.treeWidget.updateDataForNode(ownerNode, data);
-                  if (!this.isVisible) {
-                      this.delayedRefresh = true;
-                  }
-                  if(this.selectedNode === ownerNode) {
-                    this.formWidget.setSelection(this.selectedNode);
-                  }
-                  break;
-              }
-              default: { /** */}
-          }
-      });
-      this.modelServerApi.get(this.getModelIDToRequest()).then(response => {
-          if (response.statusCode === 200) {
-              if (isEqual(this.instanceData, response.body)) {
-                  return;
-              }
-              this.instanceData = response.body;
-              this.treeWidget
-                  .setData({ error: false, data: this.instanceData })
-                  .then(() => this.treeWidget.selectFirst());
-              return;
-          }
-          this.treeWidget.setData({ error: !!response.statusMessage });
-          this.renderError(
-              "An error occurred when requesting '" +
-          this.getModelIDToRequest() +
-          "' - Status " +
-          response.statusCode +
-          ' ' +
-          response.statusMessage
-          );
-          this.instanceData = undefined;
-          return;
-      });
-      this.modelServerApi.subscribe(this.getModelIDToRequest());
-      // see https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onbeforeunload
-      window.onbeforeunload = () => this.dispose();
-  }
-  private getOldSelectedPath(): string[] {
-      const paths: string[] = [];
-      if (!this.selectedNode) {
-          return paths;
-      }
-      paths.push(this.selectedNode.name);
-      let parent = this.selectedNode.parent;
-      while (parent) {
-          paths.push(parent.name);
-          parent = parent.parent;
-      }
-      paths.splice(paths.length - 1, 1);
-      return paths;
-  }
+            this.instanceData = undefined;
+            return;
+        });
+        this.modelServerApi.subscribe(this.getModelIDToRequest());
+        // see https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onbeforeunload
+        window.onbeforeunload = () => this.dispose();
+    }
+    private getOldSelectedPath(): string[] {
+        const paths: string[] = [];
+        if (!this.selectedNode) {
+            return paths;
+        }
+        paths.push(this.selectedNode.name);
+        let parent = this.selectedNode.parent;
+        while (parent) {
+            paths.push(parent.name);
+            parent = parent.parent;
+        }
+        paths.splice(paths.length - 1, 1);
+        return paths;
+    }
 
-  public save(): void {
-      this.logger.info('Save data to server');
-      this.modelServerApi.save(this.getModelIDToRequest());
-  }
+    public save(): void {
+        this.logger.info('Save data to server');
+        this.modelServerApi.save(this.getModelIDToRequest());
+    }
 
-  private getEClassFromKey(key: string): string {
-      return CoffeeModel.Type[key[0].toUpperCase() + key.slice(1)];
-  }
+    private getEClassFromKey(key: string): string {
+        return CoffeeModel.Type[key[0].toUpperCase() + key.slice(1)];
+    }
 
-  protected deleteNode(node: Readonly<TreeEditor.Node>): void {
-      const removeCommand = ModelServerCommandUtil.createRemoveCommand(
-          this.getNodeDescription(node.parent as TreeEditor.Node),
-          node.jsonforms.property,
-          node.jsonforms.index ? [Number(node.jsonforms.index)] : []
-      );
-      this.modelServerApi.edit(this.getModelIDToRequest(), removeCommand);
-  }
-  protected addNode({ node, type, property }: AddCommandProperty): void {
-      const addCommand = ModelServerCommandUtil.createAddCommand(
-          this.getNodeDescription(node),
-          property,
-          [{ eClass: type }]
-      );
-      this.modelServerApi.edit(this.getModelIDToRequest(), addCommand);
-  }
+    protected deleteNode(node: Readonly<TreeEditor.Node>): void {
+        const removeCommand = ModelServerCommandUtil.createRemoveCommand(
+            this.getNodeDescription(node.parent as TreeEditor.Node),
+            node.jsonforms.property,
+            node.jsonforms.index ? [Number(node.jsonforms.index)] : []
+        );
+        this.modelServerApi.edit(this.getModelIDToRequest(), removeCommand);
+    }
+    protected addNode({ node, type, property }: AddCommandProperty): void {
+        const addCommand = ModelServerCommandUtil.createAddCommand(
+            this.getNodeDescription(node),
+            property,
+            [{ eClass: type }]
+        );
+        this.modelServerApi.edit(this.getModelIDToRequest(), addCommand);
+    }
 
-  dispose(): void {
-      this.modelServerApi.unsubscribe(this.getModelIDToRequest());
-      super.dispose();
-  }
+    dispose(): void {
+        this.modelServerApi.unsubscribe(this.getModelIDToRequest());
+        super.dispose();
+    }
 
-  protected handleFormUpdate(data: any, node: TreeEditor.Node): void {
-      const modelServerNode = this.getNodeDescription(node);
-      Object.keys(data)
-          .filter(key => key !== 'eClass')
-          .forEach(key => {
-              if (
-                  data[key] instanceof Object &&
-          !isEqual(node.jsonforms.data[key], data[key])
-              ) {
-                  const eClass = data[key].eClass || this.getEClassFromKey(key);
-                  const setCommand = ModelServerCommandUtil.createSetCommand(
-                      modelServerNode,
-                      key,
-                      []
-                  );
-                  const toAdd = clone(data[key]);
-                  toAdd['eClass'] = eClass;
-                  setCommand.objectsToAdd = [toAdd];
-                  const ref = { eClass, $ref: '//@objectsToAdd.0' };
-                  setCommand.objectValues = [ref];
-                  this.modelServerApi.edit(this.getModelIDToRequest(), setCommand);
-              } else {
-                  const setCommand = ModelServerCommandUtil.createSetCommand(
-                      modelServerNode,
-                      key,
-                      [data[key]]
-                  );
-                  this.modelServerApi.edit(this.getModelIDToRequest(), setCommand);
-              }
-          });
-  }
+    protected handleFormUpdate(data: any, node: TreeEditor.Node): void {
+        const modelServerNode = this.getNodeDescription(node);
+        Object.keys(data)
+            .filter(key => key !== 'eClass')
+            .forEach(key => {
+                if (
+                    data[key] instanceof Object &&
+                    !isEqual(node.jsonforms.data[key], data[key])
+                ) {
+                    const eClass = data[key].eClass || this.getEClassFromKey(key);
+                    const setCommand = ModelServerCommandUtil.createSetCommand(
+                        modelServerNode,
+                        key,
+                        []
+                    );
+                    const toAdd = clone(data[key]);
+                    toAdd['eClass'] = eClass;
+                    setCommand.objectsToAdd = [toAdd];
+                    const ref = { eClass, $ref: '//@objectsToAdd.0' };
+                    setCommand.objectValues = [ref];
+                    this.modelServerApi.edit(this.getModelIDToRequest(), setCommand);
+                } else {
+                    const setCommand = ModelServerCommandUtil.createSetCommand(
+                        modelServerNode,
+                        key,
+                        [data[key]]
+                    );
+                    this.modelServerApi.edit(this.getModelIDToRequest(), setCommand);
+                }
+            });
+    }
 
-  /**
-   * Create the corresponding ModelServerReferenceDescription for the given tree node.
-   * @param node The tree node to convert
-   */
-  protected getNodeDescription(
-      node: TreeEditor.Node
-  ): ModelServerReferenceDescription {
-      const getRefSegment = (n: TreeEditor.Node): string =>
-          n.jsonforms.property
-              ? `@${n.jsonforms.property}` +
-          (n.jsonforms.index ? `.${n.jsonforms.index}` : '')
-              : '';
-      let refToNode = '';
-      let toCheck: TreeNode = node;
-      while (toCheck && TreeEditor.Node.is(toCheck)) {
-          const parentRefSeg = getRefSegment(toCheck);
-          refToNode =
-        parentRefSeg === '' ? refToNode : '/' + parentRefSeg + refToNode;
-          toCheck = toCheck.parent;
-      }
-      const ownerRef = `${
-          this.workspaceService.workspace.uri
-      }/${this.getModelIDToRequest()}#/${refToNode}`;
-      return {
-          eClass: node.jsonforms.type,
-          $ref: ownerRef.replace('file:///', 'file:/')
-      };
-  }
+    /**
+     * Create the corresponding ModelServerReferenceDescription for the given tree node.
+     * @param node The tree node to convert
+     */
+    protected getNodeDescription(
+        node: TreeEditor.Node
+    ): ModelServerReferenceDescription {
+        const getRefSegment = (n: TreeEditor.Node): string =>
+            n.jsonforms.property
+                ? `@${n.jsonforms.property}` +
+                (n.jsonforms.index ? `.${n.jsonforms.index}` : '')
+                : '';
+        let refToNode = '';
+        let toCheck: TreeNode = node;
+        while (toCheck && TreeEditor.Node.is(toCheck)) {
+            const parentRefSeg = getRefSegment(toCheck);
+            refToNode =
+                parentRefSeg === '' ? refToNode : '/' + parentRefSeg + refToNode;
+            toCheck = toCheck.parent;
+        }
+        const ownerRef = `${this.workspaceService.workspace.uri
+            }/${this.getModelIDToRequest()}#/${refToNode}`;
+        return {
+            eClass: node.jsonforms.type,
+            $ref: ownerRef.replace('file:///', 'file:/')
+        };
+    }
 
-  private getModelIDToRequest(): string {
-      const rootUriLength = this.workspaceService
-          .getWorkspaceRootUri(this.options.uri)
-          .toString().length;
-      return this.options.uri.toString().substring(rootUriLength + 1);
-  }
+    private getModelIDToRequest(): string {
+        const rootUriLength = this.workspaceService
+            .getWorkspaceRootUri(this.options.uri)
+            .toString().length;
+        return this.options.uri.toString().substring(rootUriLength + 1);
+    }
 
-  protected configureTitle(title: Title<Widget>): void {
-      title.label = this.options.uri.path.base;
-      title.caption = BaseTreeEditorWidget.WIDGET_LABEL;
-      title.closable = true;
-      title.iconClass = 'fa coffee-icon dark-purple';
-  }
+    protected configureTitle(title: Title<Widget>): void {
+        title.label = this.options.uri.path.base;
+        title.caption = BaseTreeEditorWidget.WIDGET_LABEL;
+        title.closable = true;
+        title.iconClass = 'fa coffee-icon dark-purple';
+    }
 
-  show(): void {
-      super.show();
-      if (this.delayedRefresh) {
-          this.delayedRefresh = false;
-          this.treeWidget.model.refresh();
-      }
-  }
+    show(): void {
+        super.show();
+        if (this.delayedRefresh) {
+            this.delayedRefresh = false;
+            this.treeWidget.model.refresh();
+        }
+    }
 }
 export namespace CoffeeTreeEditorConstants {
-  export const WIDGET_ID = 'json-forms-tree-editor';
-  export const EDITOR_ID = 'com.eclipsesource.coffee.editor';
+    export const WIDGET_ID = 'json-forms-tree-editor';
+    export const EDITOR_ID = 'com.eclipsesource.coffee.editor';
 }

--- a/web/coffee-server/src/browser/coffee-frontend-contribution.ts
+++ b/web/coffee-server/src/browser/coffee-frontend-contribution.ts
@@ -1,0 +1,26 @@
+/********************************************************************************
+ * Copyright (c) 2021 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * https://www.eclipse.org/legal/epl-2.0, or the MIT License which is
+ * available at https://opensource.org/licenses/MIT.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR MIT
+ ********************************************************************************/
+import { ModelServerFrontendContribution } from '@eclipse-emfcloud/modelserver-theia/lib/browser';
+import { injectable } from 'inversify';
+
+@injectable()
+export class CoffeeModelServerFrontendContribution extends ModelServerFrontendContribution {
+
+    // #FIXME: if theia is updated to at least 1.5.0 this contribution can be removed again
+    async setup(): Promise<void> {
+        this.workspaceService.onWorkspaceChanged(workspace => {
+            if (workspace[0] && workspace[0].uri) {
+                const workspaceRoot = workspace[0].uri.toString();
+                this.modelServerClient.configure({ workspaceRoot });
+            }
+        });
+    }
+}

--- a/web/coffee-server/src/browser/coffee-server-frontend-module.ts
+++ b/web/coffee-server/src/browser/coffee-server-frontend-module.ts
@@ -20,6 +20,7 @@ import {
 } from '@theia/core/lib/browser/frontend-application';
 import { ContainerModule, inject, injectable } from 'inversify';
 
+import { CoffeeModelServerFrontendContribution } from './coffee-frontend-contribution';
 import { TimeoutDialog, TimeoutDialogProps } from './coffee-timeout-dialog';
 
 export default new ContainerModule(bind => {
@@ -27,7 +28,8 @@ export default new ContainerModule(bind => {
     bind(TimeoutDialog).toSelf().inSingletonScope();
     bind(BackendWatcher).toSelf().inSingletonScope();
     bind(FrontendApplicationContribution).toService(BackendWatcher);
-
+    bind(CoffeeModelServerFrontendContribution).toSelf().inSingletonScope();
+    bind(FrontendApplicationContribution).toService(CoffeeModelServerFrontendContribution);
 });
 
 @injectable()

--- a/web/coffee-server/src/node/backend-module.ts
+++ b/web/coffee-server/src/node/backend-module.ts
@@ -17,17 +17,16 @@ import { GLSPLaunchOptions, GLSPServerLauncher } from './glsp-server-launcher';
 
 export default new ContainerModule((bind, _unbind, isBound, rebind) => {
     if (isBound(LaunchOptions)) {
-        rebind(LaunchOptions).to(CoffeeLaunchOptions).inSingletonScope();
+        rebind(LaunchOptions).to(CoffeeModelServerLaunchOptions).inSingletonScope();
     } else {
-        bind(LaunchOptions).to(CoffeeLaunchOptions).inSingletonScope();
+        bind(LaunchOptions).to(CoffeeModelServerLaunchOptions).inSingletonScope();
     }
     bind(GLSPLaunchOptions).to(CoffeeGlspLaunchOptions).inSingletonScope();
     bind(BackendApplicationContribution).to(GLSPServerLauncher);
 });
 
 @injectable()
-export class CoffeeLaunchOptions implements LaunchOptions {
-    isRunning = false;
+export class CoffeeModelServerLaunchOptions implements LaunchOptions {
     baseURL = 'api/v1/';
     serverPort = 8081;
     hostname = 'localhost';
@@ -37,8 +36,8 @@ export class CoffeeLaunchOptions implements LaunchOptions {
 
 @injectable()
 export class CoffeeGlspLaunchOptions implements GLSPLaunchOptions {
-    isRunning: false;
-    serverPort: 5008;
-    hostname: 'localhost';
+    isRunning = false;
+    serverPort = 5008;
+    hostname = 'localhost';
     jarPath = resolve(join(__dirname, '..', '..', 'build', 'com.eclipsesource.workflow.glsp.server-0.0.1-SNAPSHOT-glsp.jar'));
 }

--- a/web/coffee-server/src/node/glsp-server-launcher.ts
+++ b/web/coffee-server/src/node/glsp-server-launcher.ts
@@ -88,13 +88,13 @@ export class GLSPServerLauncher implements BackendApplicationContribution {
 
     protected logError(data: string | Buffer): void {
         if (data) {
-            this.logger.error(`ModelServerBackendContribution: ${data}`);
+            this.logger.error(`GLSPServerLauncher: ${data}`);
         }
     }
 
     protected logInfo(data: string | Buffer): void {
         if (data) {
-            this.logger.info(`ModelServerBackendContribution: ${data}`);
+            this.logger.info(`GLSPServerLauncher: ${data}`);
         }
     }
 

--- a/web/coffee-workflow-analyzer-editors/src/node/workflow-lsp-launcher.ts
+++ b/web/coffee-workflow-analyzer-editors/src/node/workflow-lsp-launcher.ts
@@ -72,13 +72,13 @@ export class WorkflowLSPServerLauncher implements BackendApplicationContribution
 
     protected logError(data: string | Buffer): void {
         if (data) {
-            this.logger.error(`ModelServerBackendContribution: ${data}`);
+            this.logger.error(`WorkflowLSPServerLauncher: ${data}`);
         }
     }
 
     protected logInfo(data: string | Buffer): void {
         if (data) {
-            this.logger.info(`ModelServerBackendContribution: ${data}`);
+            this.logger.info(`WorkflowLSPServerLauncher: ${data}`);
         }
     }
 

--- a/web/coffee-workflow-glsp-editor/src/node/backend-module.ts
+++ b/web/coffee-workflow-glsp-editor/src/node/backend-module.ts
@@ -13,24 +13,11 @@
  *
  * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
  ********************************************************************************/
-import { LaunchOptions } from '@eclipse-emfcloud/modelserver-theia';
 import { LanguageServerContribution } from '@theia/languages/lib/node';
-import { ContainerModule, injectable } from 'inversify';
-import { join, resolve } from 'path';
+import { ContainerModule } from 'inversify';
 
 import { WorkflowNotationGLServerContribution } from './workflow-glsp-server-contribution';
 
 export default new ContainerModule(bind => {
     bind(LanguageServerContribution).to(WorkflowNotationGLServerContribution).inSingletonScope();
-    bind(LaunchOptions).to(ModelServerLaunchOptions).inSingletonScope();
 });
-
-@injectable()
-export class ModelServerLaunchOptions implements LaunchOptions {
-    isRunning = false;
-    baseURL = 'api/v1/';
-    serverPort = 8081;
-    hostname = 'localhost';
-    jarPath = resolve(join(__dirname, '..', '..', 'build', 'com.eclipsesource.modelserver.example-0.0.1-SNAPSHOT-standalone.jar'));
-    additionalArgs = ['--errorsOnly'];
-}

--- a/web/package.json
+++ b/web/package.json
@@ -56,6 +56,7 @@
     "**/@eclipse-glsp/client": "0.8.0-rc-02",
     "**/@eclipse-glsp/theia-integration": "0.8.0-rc-02",
     "**/@eclipse-glsp-examples/workflow-sprotty": "0.8.0-rc-02",
+    "**/@eclipse-emfcloud/modelserver-theia": "0.8.0-next.5bf1df0",
     "**/typescript": "^3.9.3"
   }
 }

